src/App.js

```diff
@@ -19,64 +20,47 @@ class App extends Component {
         text: 'FPO image',
       },
       { type: BlockType.Blockquote, text: 'this is a quote' },
-    ].map(this.toObjectsWithKeys),
+    ].map(this.toObjectsWithKey),
   };
 
-  toObjectsWithKeys(block, index) {
+  toObjectsWithKey(block, index) {
     const key = btoa(`${index}: ${JSON.stringify(block)}`);
     return { key, ...block };
   }
 
-  // step 2: refactor the addtype functions into generic add block function where you just pass in the block.type as an argument
-  addType = option => {
-    option = BlockType;
-    const newOption = this.toObjectsWithKeys(
-      { type: BlockType.Heading, text: 'testing' },
+  addBlock = type => {
+    const newBlock = this.toObjectsWithKey(
+      { type, text: `${new Date().toLocaleString()}` },
       this.state.blocks.length,
     );
-    this.setState({ blocks: [...this.state.blocks, newOption] });
+    if (type === BlockType.Image) {
+      newBlock.url = 'http://fpoimg.com/300x200';
+    }
+    this.setState({ blocks: [...this.state.blocks, newBlock] });
   };
 
   addHeading = () => {
-    const newHeading = this.toObjectsWithKeys(
-      { type: BlockType.Heading, text: 'testing' },
-      this.state.blocks.length,
-    );
-    this.setState({ blocks: [...this.state.blocks, newHeading] });
+    this.addBlock(BlockType.Heading);
   };
 
   addText = () => {
-    const newText = this.toObjectsWithKeys(
-      { type: BlockType.Text, text: 'Test number 2 on the button' },
-      this.state.blocks.length,
-    );
-    this.setState({ blocks: [...this.state.blocks, newText] });
+    this.addBlock(BlockType.Text);
   };
 
-  addImage = () => {
-    const newImage = this.toObjectsWithKeys(
-      {
-        type: BlockType.Image,
-        url: 'http://fpoimg.com/300x200',
-        text: 'FPO image',
-      },
-      this.state.blocks.length,
-    );
-    this.setState({ blocks: [...this.state.blocks, newImage] });
+  addBlockquote = () => {
+    this.addBlock(BlockType.Blockquote);
   };
 
-  addBlockquote = () => {
-    const newBlockquote = this.toObjectsWithKeys(
-      { type: BlockType.Blockquote, text: 'this is a NEW quote' },
-      this.state.blocks.length,
-    );
-    this.setState({ blocks: [...this.state.blocks, newBlockquote] });
+  addImage = () => {
+    this.addBlock(BlockType.Image);
   };
 
-  toComponents(block) {
+  //step 3.1, make custom components for other return statment of toComponent function.
+
+  toComponent(block) {
     switch (block.type) {
       case BlockType.Heading:
-        return <h2 key={block.key}>{block.text}</h2>;
+        return <Heading key={block.key} text={block.text} />;
```

Comment:
Just FYI, another way you might see this done out in the wild is like `<Heading { ...block } />` this is called the "spread props" style. I just want you to be aware of it. I actually kind of hate it and think it's bad because you wind up passing a bunch of props you don't need or mean to down.

When your component hierarchy gets really deep you wind up not knowing what's getting passed in to the "leaf nodes" at the very end. This has been the cause of a bunch of weird behavior for me in production and it's a pain to track down. Anyway:

![giphy](https://user-images.githubusercontent.com/810398/55043487-e5fe7900-500c-11e9-9071-8c25b9bccba0.gif)

Reply:
I think I understand this @mysterycommand but can we just touch base super quick on the call tomorrow?

Reply:
I actually think we're doing this in the toObjectWithKey function on lines 26-28.